### PR TITLE
br: restore merge split ranges with small data

### LINKED ISSUE #53532: br restore splits too much regions if there are too many small tables whose data size is less than region size.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. create so many tables, but each table only has a little data.
2. backup cluster with `split-table=false/split-region-on-table=false`
3. restore cluster

### 2. What did you expect to see? (Required)
the number of regions of downstream cluster is closed to the number of upstream cluster.
### 3. What did you see instead (Required)
the number of regions has many regions.
### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by Leavrth**
### What problem does this PR solve?

Issue Number: close #53532 

Problem Summary:
br restore splits too many regions if there are too many small tables whose data size is less than region size.

### What changed and how does it work?
ignore some keys to split less regions
Due to https://github.com/tikv/tikv/issues/17278, BR needs merge the generated `SSTMeta` at the same region into one `Ingest` request.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/conn/conn.go [hunk 1]
```diff
@@ -324,6 +325,14 @@ func (mgr *Mgr) ProcessTiKVConfigs(ctx context.Context, cfg *kvconfig.KVConfig,
 		if err != nil {
 			return err
 		}
+		if !splitRegionOnTable.Modified {
```
Comment (3pointer) [inline br/pkg/conn/conn.go:328]: This seems always be true. so no need check this.

File: br/pkg/conn/conn.go [hunk 2]
```diff
@@ -324,6 +325,14 @@ func (mgr *Mgr) ProcessTiKVConfigs(ctx context.Context, cfg *kvconfig.KVConfig,
...
+			splitTable, e := kvconfig.ParseSplitRegionOnTable(respBytes)
+			if e != nil {
+				log.Warn("Failed to parse split region on table from config", logutil.ShortError(e))
+				return e
+			}
+			splitRegionOnTable.Value = splitRegionOnTable.Value || splitTable
```
Comment (3pointer) [inline br/pkg/conn/conn.go:334]: It's better to check all tikv set split-on-table to true here.

File: br/pkg/restore/snap_client/import.go [hunk 3]
```diff
@@ -565,29 +551,34 @@ func (importer *SnapFileImporter) buildDownloadRequest(
...
 	downloadMetasMap := make(map[string]import_sstpb.SSTMeta)
 	resultMetasMap := make(map[string]*import_sstpb.SSTMeta)
 	downloadReqsMap := make(map[string]*import_sstpb.DownloadRequest)
-	for _, file := range files {
-		req, sstMeta, err := importer.buildDownloadRequest(file, rewriteRules, regionInfo, cipher)
-		if err != nil {
-			return nil, errors.Trace(err)
+	for _, files := range filesGroup {
+		for _, file := range files.Files {
+			req, sstMeta, skip, err := importer.buildDownloadRequest(file, files.RewriteRules, regionInfo, cipher)
```
Comment (3pointer) [inline br/pkg/restore/snap_client/import.go:570]: It's better to use `req !=nil` to get rid of `skip` return  here.

Comment (Leavrth) [reply br/pkg/restore/snap_client/import.go]: done

**Review by YuJuncen — COMMENTED** [review]
rest lgtm

**Review by YuJuncen — APPROVED** [review]

File: br/pkg/restore/snap_client/tikv_sender.go [hunk 4]
```diff
@@ -119,16 +179,103 @@ func SortAndValidateFileRanges(
...
+			// split key generation
+			afterMergedGroupSize := groupSize + rg.Size
+			afterMergedGroupCount := groupCount + rg.Count
+			if afterMergedGroupSize > splitSizeBytes || afterMergedGroupCount > splitKeyCount || mergedRangeCount > MergedRangeCountThreshold {
+				log.Info("merge ranges across tables due to kv size/count or merged count threshold exceeded",
+					zap.Uint64("merged kv size", groupSize),
+					zap.Uint64("merged kv count", groupCount),
+					zap.Int("merged range count", mergedRangeCount))
+				groupSize, groupCount = rg.Size, rg.Count
+				mergedRangeCount = 0
```
Comment (3pointer) [inline br/pkg/restore/snap_client/tikv_sender.go:213]: shall we reset this value after `lastFilesGroup` appended?

Comment (Leavrth) [reply br/pkg/restore/snap_client/tikv_sender.go:213]: `mergedRangeCount` records the number of files in `tableIDWithFilesGroup` to make sure the batch size of ingest is not too large.

**Review by 3pointer — APPROVED** [review]

**Pull request merged**